### PR TITLE
Key `subscriptions.cache` by user account ID

### LINKED ISSUE #3654: [WebToolsE2E][AspireDeployment] When multiple user accounts are signed in, publish Aspire Starter App to ACA using VS fails with failed to resolve user access to subscription
**INSTALL STEPS**
1. Clean machine: Win11 x64 23h2 ENU
2 Install VS 17.10 Preview 3 [latest build, includes Aspire 8.0.0-preview.5.24201.12]
- Note: you need a build that includes Antonio's fix for AzDO bug 2016615. Otherwise, you will run into the issue in that bug first
- Apply NuGet Feeds
3. Install latest daily azd build
- I got azd version 1.8.0-beta.1-daily.3665690

**REPRO STEPS**
1. Open VS, sign-in to 2 accounts in Account Settings
- I signed in with both my work account (@microsoft.com) as well as a live.com account (@live.com). Both have access to their respective Azure subscriptions.

**MVC app - build, run and publish using live account**
2. New ASP.NET Core MVC web app > WebAppMvc2
3. Build and run, then publish to Azure App Service (Windows) using the live.com account - this works fine.
4. Close the solution

**Aspire Starter app - build, run and publish using live account**
5. New Aspire Starter app > Use redis option is unchecked > AspireStarterApp2
6. Build and run, 
7. Publish to ACA - create a new publish environment using the live.com account
8. Hit Publish in VS

**ACTUAL**
Publish fails
image

```
Build started at 5:57 PM...
1>------ Build started: Project: AspireStarterApp2.ServiceDefaults, Configuration: Debug Any CPU ------
1>C:\Program Files\dotnet\sdk\8.0.300-preview.24170.4\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.targets(311,5): message NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy
1>AspireStarterApp2.ServiceDefaults -> C:\Users\balach\source\repos\AspireStarterApp2\AspireStarterApp2.ServiceDefaults\bin\Debug\net8.0\AspireStarterApp2.ServiceDefaults.dll
2>------ Build started: Project: AspireStarterApp2.Web, Configuration: Debug Any CPU ------
3>------ Build started: Project: AspireStarterApp2.ApiService, Configuration: Debug Any CPU ------
3>C:\Program Files\dotnet\sdk\8.0.300-preview.24170.4\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.targets(311,5): message NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy
2>AspireStarterApp2.Web -> C:\Users\balach\source\repos\AspireStarterApp2\AspireStarterApp2.Web\bin\Debug\net8.0\AspireStarterApp2.Web.dll
3>AspireStarterApp2.ApiService -> C:\Users\balach\source\repos\AspireStarterApp2\AspireStarterApp2.ApiService\bin\Debug\net8.0\AspireStarterApp2.ApiService.dll
4>------ Build started: Project: AspireStarterApp2.AppHost, Configuration: Debug Any CPU ------
4>AspireStarterApp2.AppHost -> C:\Users\balach\source\repos\AspireStarterApp2\AspireStarterApp2.AppHost\bin\Debug\net8.0\AspireStarterApp2.AppHost.dll
========== Build: 4 succeeded, 0 failed, 0 up-to-date, 0 skipped ==========
========== Build completed at 5:57 PM and took 05.947 seconds ==========
4/4/2024 5:57:48 PM: Info 
4/4/2024 5:57:48 PM: Info Provisioning Azure resources (azd provision)
4/4/2024 5:57:48 PM: Info Provisioning Azure resources can take some time.
4/4/2024 5:57:48 PM: Info 
4/4/2024 5:57:49 PM: Info Analyzing Aspire Application (this might take a moment...)
4/4/2024 5:57:56 PM: Info Initialize bicep provider
4/4/2024 5:57:59 PM: Info Creating a deployment plan
4/4/2024 5:57:59 PM: Error deployment failed: failing invoking action 'provision', error deploying infrastructure: resolving bicep parameters file: fetching current principal id: getting tenant id for subscription xyz. Error: failed to resolve user access to subscription with ID 'xyz'. If you recently gained access to this subscription, run `azd auth login` again to reload subscriptions.
Otherwise, visit this subscription in Azure Portal using the browser, then run `azd auth login` 
```

**Workaround**
Open Command Prompt, run `azd auth logout`. Then go back to VS and hit 'Publish' again. It works fine this time.

**EXPECTED**
Publish to succeed

### PR DESCRIPTION
**Pull request by weikanglim**
Key `subscriptions.cache` by user account ID.

The cache now stores a map of account ID to subscriptions. By default, account ID is chosen to be the `oid` claim of the user. When `oid` claim is not present (personal accounts), the `sub` (subject) claim is used as the account ID instead. 

This allows VS users to be able to switch between accounts without require running `azd auth logout` in-between.

Manually tested switching between personal and work accounts in CLI and VS.

Fixes #3654, #2301

### CODE REVIEW

File: cli/azd/pkg/auth/manager.go [hunk 1]
```diff
@@ -332,6 +332,36 @@ func (m *Manager) CredentialForCurrentUser(
 	return nil, ErrNoCurrentUser
 }
 
+type ClaimsForCurrentUserOptions = CredentialForCurrentUserOptions
```
Comment (weikanglim) [inline cli/azd/pkg/auth/manager.go:335]: Please review `auth` changes separately as #3843

File: cli/azd/pkg/account/subscriptions_cache.go [hunk 2]
```diff
@@ -12,88 +14,116 @@ import (
...
+	// load cache from disk
+	cacheFile, err := os.ReadFile(filepath.Join(s.cacheDir, cSubscriptionsCacheFile))
 	if err != nil {
 		return nil, err
 	}
 
-	var subscriptions []Subscription
-	err = json.Unmarshal(cacheFile, &subscriptions)
+	var cache map[string][]Subscription
+	err = json.Unmarshal(cacheFile, &cache)
```
Comment (ellismg) [inline cli/azd/pkg/account/subscriptions_cache.go:65]: If the `subscriptions.cache` file is the older version, the unmarshal here will fail, correct?  I see in `Save` we ignore this error and plow ahead (reasonable), do we need to do something here as well or does a larger part of the system handle the fact that `Get` may not return an error even when there are not I/O errors?

Comment (ellismg) [reply cli/azd/pkg/account/subscriptions_cache.go:65]: I see - `GetSubscriptions` is the only caller and on an any error it just ignores the cache and does a full fetch, so this feels good.

**Review by ellismg — APPROVED** [review]
LGTM - Thanks for the fix and also teasing apart the PR :-).

File: cli/azd/pkg/account/subscriptions_cache.go (hunk 2 above)
Comment (weikanglim) [reply cli/azd/pkg/account/subscriptions_cache.go:65]: Made this a non-exported type as  `subscriptionsCache` to make the helper/internal component clearer.

**Pull request merged**